Hydro/src/Hydro/Renderer/Renderer2D.cpp

```diff
@@ -112,49 +154,61 @@ namespace Hydro
 
 	void Renderer2D::ShutDown()
 	{
-		s_Data->QuadDescriptorSet->Destory();
-		s_Data->VulkanTexture->Destory();
-		s_Data->Shader->Destory();
-		s_Data->QuadPipeline->ShutDown();
-		s_Data->QuadUniformBuffer->Destory();
-		s_Data->QuadIndexBuffer->Destory();
-		s_Data->QuadVertexBuffer->Destory();
+		s_Data->QuadDescriptorSet->Destroy();
+		s_Data->VulkanTexture->Destroy();
+		s_Data->Shader->Destroy();
+		s_Data->QuadPipeline->Destroy();
+		s_Data->QuadUniformBuffer->Destroy();
+		s_Data->QuadIndexBuffer->Destroy();
+		s_Data->QuadVertexBuffer->Destroy();
 
 		s_Data = nullptr;
 	}
 
 	void Renderer2D::Begin()
 	{
-		uint32_t currentImage = Renderer::GetRenderFrame();
-		auto& extent = Renderer::GetVulkanSwapChain()->GetExtend();
-
-		static auto startTime = std::chrono::high_resolution_clock::now();
+		s_Data->QuadIndexCount = 0;
+		s_Data->QuadVertexBufferPtr = s_Data->QuadVertexBufferBase;
 
-		auto currentTime = std::chrono::high_resolution_clock::now();
-		float time = std::chrono::duration<float, std::chrono::seconds::period>(currentTime - startTime).count();
-
-		glm::mat4 model = glm::rotate(glm::mat4(1.0f), time * glm::radians(90.0f), glm::vec3(0.0f, 0.0f, 1.0f));
-		glm::mat4 view = glm::lookAt(glm::vec3(2.0f, 2.0f, 2.0f), glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 0.0f, 1.0f));
-		glm::mat4 proj = glm::perspective(glm::radians(45.0f), extent.width / (float)extent.height, 0.1f, 10.0f);
-		proj[1][1] *= -1;
+		uint32_t currentImage = Renderer::GetRenderFrame();
 
-		s_Data->ubo.mvp = proj * view * model;
-		s_Data->QuadUniformBuffer->Update(&s_Data->ubo.mvp, currentImage, (uint8_t)sizeof(UniformBufferObject));
+		s_Data->QuadUniformBuffer->Update(&s_Data->camera->GetViewProjectMatrix(), currentImage, (uint8_t)sizeof(UniformBufferObject));
 	}
 
 	void Renderer2D::End()
-	{
-	}
-
-	void Renderer2D::DrawQuad()
 	{
 		auto commandBuffer = Renderer::GetVulkanSwapChain()->GetCommandBuffer();
-
 		s_Data->QuadPipeline->Bind();
 		s_Data->QuadVertexBuffer->Bind();
 		s_Data->QuadIndexBuffer->Bind();
 		s_Data->QuadPipeline->BindDescriptorSets();
 
-		vkCmdDrawIndexed(commandBuffer, static_cast<uint32_t>(s_Data->indices.size()), 1, 0, 0, 0);
+		uint32_t dataSize = (uint8_t*)s_Data->QuadVertexBufferPtr - (uint8_t*)s_Data->QuadVertexBufferBase;
+		s_Data->QuadVertexBuffer->SetData(s_Data->QuadVertexBufferBase, dataSize);
+
+		vkCmdDrawIndexed(commandBuffer, static_cast<uint32_t>(s_Data->QuadIndexCount), 1, 0, 0, 0);
+	}
+
+	void Renderer2D::DrawQuad(const glm::mat4& transform, const glm::vec4& color)
+	{
+		constexpr size_t quadVertexCount = 4;
+		const float textureIndex = 0.0f; // White Texture
+		constexpr glm::vec2 textureCoords[] = { { 0.0f, 0.0f }, { 1.0f, 0.0f }, { 1.0f, 1.0f }, { 0.0f, 1.0f } };
+		const float tilingFactor = 1.0f;
+
+		// if (s_Data->QuadIndexCount >= Renderer2DData::MaxIndices)
```

Comment:
Still needs to be done. (Create new empty buffers to render if we reach the limit) (render to screen create new ones for next batch)